### PR TITLE
perf(diffs): parallelize diffs

### PR DESCRIPTION
**Pull request by JonasBa**
Introduce parallelism for diffing images

We were currently processing images using only a single CPU core. This change introduces the notion of parallelism by using a worker pool and dispatching actions to the available workers (threads). Our [ubuntu runner](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) currently has 2 CPU cores so by using an extra thread, I'm hoping we should see a linear performance improvement.

I have tested running the process on macos runner since it has 3 CPU cores, but the results were worse than using ubuntu (+30% or about 2 minutes compared to ubuntu over a sample of 5 runs). There is an open issue that is worth monitoring as this seems to be a common problem https://github.com/actions/virtual-environments/issues/1336. I have not investigated where the performance decrease actually comes from.

Running some tests with the current changes, we seem to be consistently able to run a full diff (~900 snapshots) in about 4 minutes. Our current p50 over last 90 days is at [7 minutes](https://sentry.io/organizations/sentry/performance/?project=5324467&statsPeriod=90d) which would roughly match my expectations of linear improvements as we add more cores.

Note: Please dismiss the large diff, I had to rebuild the index.js and added the separately compiled worker in order to be able to test this.

A few of the full test runs (there was little difference for cases where early termination was triggered):
- [ubuntu runner with parallelism 2](https://github.com/getsentry/sentry/runs/6777427887?check_suite_focus=true): 4min27s
- [macos runner with parallelism 3](https://github.com/getsentry/sentry/runs/6748958858?check_suite_focus=true): 7min39s

### CODE REVIEW

File: src/util/diffSnapshots.ts [hunk 1]
```diff
@@ -170,81 +177,139 @@ export async function diffSnapshots({
...
 
-        missingSnapshots.delete(file);
-      } catch (err) {
-        core.debug(`Unable to diff: ${err.message}`);
-        throw err;
-      }
+      promises.push(promise);
     } else {
-      newSnapshots.add(file);
+      // If there is nothing to diff, return a
```
Comment (billyvg) [inline src/util/diffSnapshots.ts:282]: Missed some words

**Review by billyvg — COMMENTED** [review]
Adds some complexity, but can't argue against the results! 👏 👏

File: src/main.ts [hunk 2]
```diff
@@ -23,6 +24,7 @@ import {Await} from './types';
 import {getPixelmatchOptions} from './getPixelmatchOptions';
 import {downloadOtherWorkflowArtifact} from './api/downloadOtherWorkflowArtifact';
 
+const parallelism = os.cpus().length;
```
[the remark this answers is not in the document]
Comment (JonasBa) [reply src/main.ts:27]: Good call

File: src/util/diffSnapshots.ts [hunk 3]
```diff
@@ -78,8 +79,14 @@ export async function diffSnapshots({
...
   missingDirName = 'missing',
   pixelmatchOptions,
+  parallelism,
   maxChangedSnapshots = DEFAULT_MAX_CHANGED_SNAPSHOTS,
 }: DiffSnapshotsParams) {
+  if (parallelism < 1 || parallelism > os.cpus().length) {
+    throw new Error(
+      'Parallelism option should be in the range of 1 - CPU count'
+    );
+  }
```
[the remark this answers is not in the document]
Comment (JonasBa) [reply src/util/diffSnapshots.ts]: @billyvg no improvements. I think it just hogs the CPU and ends up competing for resources - 
https://github.com/getsentry/sentry/runs/6779284856?check_suite_focus=true

**Pull request merged**